### ISSUE TITLE
[new feature request] Base64 encode images into html

### ISSUE THREAD

**Issue by Postur** [manuth/MarkdownConverter#102]
Would be nice to insert the images as Base64 encoded strings.
so that everything is contained in the html file.

**Comment by manuth** [comment]
Hello @Postur 
I did a quick research and sadly there is no markdown-it plugin that suits your need.
Searching the DOM-tree for `img`-tags would be quite a butthurt, too, as there isn't really a standardized way to do so.

You could change the `markdownConverter.ConversionType` to `SelfContainedHTML` which creates a self-contained directory which contains all pictures as well as other resources.

Alternatively you could use base64-links in your markdown-document itself:

```md
![](data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAADAAAAAwCAYAAABXAvmHAAAABGdBTUEAALGPC/xhBQAAACBjSFJNAAB6JgAAgIQAAPoAAACA6AAAdTAAAOpgAAA6mAAAF3CculE8AAAABmJLR0QAAAAAAAD5Q7t/AAAACXBIWXMAAG66AABuugHW3rEXAAAAB3RJTUUH4QIKFjcfARWjgAAAAuVJREFUaN7tmT1IHEEYhtdfBOMpiEjujgtWNoKdrU0EMfGiRYKFVkJiOBU0FiEmhZWVpWACJ1EQWxEEISb+XaGFxEiMqJWFVgmIcJ16eQZm4Thcd2939ifhBp7GO9f3kZ35Zr7RtHvG6/DHFDyDIs2H8S42J4hDytYDCJ6Ba1iFdq9ECFwE7bAK15BxIpDNGnS4JSKDd8CaCJ2NKgGdb/AUihUFL4Yn8DU3uFsCOuvQaVdEBu+EdaPgbgvobMjJXmIxeImcnBtmwb0S0NmCLiMRGbwLtqwG91pAZxu6E+FPpeI5b2OfSwnRDdv5BvdLQCc1Ek2+cRLcV4EBQCAz7jB8QeB/FrhVJHDrl4DYWgzCqU2BUxi8a+vglcCS/F4NDMGxRYFjGIIaWR+W/BJYzvl+NSTgyEDgCBJQnVPglgMhkPV7IRhA4FAK/CLEAIQMKnSwBPTx8uFM1Wg0+fjDo/kqk71RMAXy2JUWBAoCBYGCQEEgmAJ9sOOzwA702X44IcuhB3Y9FtiFHihX0jUzEHFDQF1wAjbcI7IHK6CigbsCe0bB+VmDXYE/MAn1d3xWAc1O+6SyH9oMFXd8Vg+T8NvpKnQmDyyVHnWnK+WB50z1MroPL1Q1dQ16pc9h3+06IM7CrYrDt8IX1XXg5z0S4uJjAZocBm+CBf0iw4BDuwK1MGtSja9gCiJ5Bo/AFFyZVOFZqHW6nA5D2kTkHMbEedgkeAjG4NwkeBqGlb2j4p2HEwv7o0PohdKc4KJD3SteBwuduBMxJ5SvEuI1EdXXYmtxE9pGoskywrTBpsU2oqjKEdfWaPGfhQk5ic16ozcIfB+Pzd1YCC4m8QSUeXKHK29iLhT1Ri/EzY3nF9GEbJS3MU4ExOVHo+bXENsLmLYpMA0PtCAMAvfDpUWBS+jXgjYI3QIHJgIH0KIFdRC8DhYNBBahTgv6EGcEUZURSEuBtKy+Rdq/NF6FZ+Kj0eSP97H5uFt/4y9GA721POPJpAAAAABJRU5ErkJggg==)
```

Does that suit your need?

**Comment by Postur** [comment]
I took this from [yzhang-gh/vscode-markdown/src/print.ts](https://github.com/yzhang-gh/vscode-markdown/blob/3cb5ec58747b773d114ba180cf7087ceb17244cc/src/print.ts#L95)

maybe something like this could be implemented?
even works with svg files.

```
    if (configToBase64) {
        body = body.replace(imgTagRegex, function (_, p1, p2, p3) {
            if (p2.startsWith('http') || p2.startsWith('data:')) {
                return _;
            }

            const imgSrc = relToAbsPath(doc.uri, p2);
            try {
                let imgExt = path.extname(imgSrc).slice(1);
                if (imgExt === "jpg") {
                    imgExt = "jpeg";
                } else if (imgExt === "svg") {
                    imgExt += "+xml";
                }
                const file = fs.readFileSync(imgSrc.replace(/%20/g, '\ ')).toString('base64');
                return `${p1}data:image/${imgExt};base64,${file}${p3}`;
            } catch (e) {
                window.showWarningMessage(localize("ui.general.messageUnableToReadFile", imgSrc) + ` ${localize("ui.exporting.messageRevertingToImagePaths")} (${doc.fileName})`);
            }

            if (configAbsPath) {
                return `${p1}file:///${imgSrc}${p3}`;
            } else {
                return _;
            }
        });
```

**Comment by Postur** [comment]
> Alternatively you could use base64-links in your markdown-document itself:
> 
> ```lisp
> ![](data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAADAAAAAwCAYAAABXAvmHAAAABGdBTUEAALGPC/xhBQAAACBjSFJNAAB6JgAAgIQAAPoAAACA6AAAdTAAAOpgAAA6mAAAF3CculE8AAAABmJLR0QAAAAAAAD5Q7t/AAAACXBIWXMAAG66AABuugHW3rEXAAAAB3RJTUUH4QIKFjcfARWjgAAAAuVJREFUaN7tmT1IHEEYhtdfBOMpiEjujgtWNoKdrU0EMfGiRYKFVkJiOBU0FiEmhZWVpWACJ1EQWxEEISb+XaGFxEiMqJWFVgmIcJ16eQZm4Thcd2939ifhBp7GO9f3kZ35Zr7RtHvG6/DHFDyDIs2H8S42J4hDytYDCJ6Ba1iFdq9ECFwE7bAK15BxIpDNGnS4JSKDd8CaCJ2NKgGdb/AUihUFL4Yn8DU3uFsCOuvQaVdEBu+EdaPgbgvobMjJXmIxeImcnBtmwb0S0NmCLiMRGbwLtqwG91pAZxu6E+FPpeI5b2OfSwnRDdv5BvdLQCc1Ek2+cRLcV4EBQCAz7jB8QeB/FrhVJHDrl4DYWgzCqU2BUxi8a+vglcCS/F4NDMGxRYFjGIIaWR+W/BJYzvl+NSTgyEDgCBJQnVPglgMhkPV7IRhA4FAK/CLEAIQMKnSwBPTx8uFM1Wg0+fjDo/kqk71RMAXy2JUWBAoCBYGCQEEgmAJ9sOOzwA702X44IcuhB3Y9FtiFHihX0jUzEHFDQF1wAjbcI7IHK6CigbsCe0bB+VmDXYE/MAn1d3xWAc1O+6SyH9oMFXd8Vg+T8NvpKnQmDyyVHnWnK+WB50z1MroPL1Q1dQ16pc9h3+06IM7CrYrDt8IX1XXg5z0S4uJjAZocBm+CBf0iw4BDuwK1MGtSja9gCiJ5Bo/AFFyZVOFZqHW6nA5D2kTkHMbEedgkeAjG4NwkeBqGlb2j4p2HEwv7o0PohdKc4KJD3SteBwuduBMxJ5SvEuI1EdXXYmtxE9pGoskywrTBpsU2oqjKEdfWaPGfhQk5ic16ozcIfB+Pzd1YCC4m8QSUeXKHK29iLhT1Ri/EzY3nF9GEbJS3MU4ExOVHo+bXENsLmLYpMA0PtCAMAvfDpUWBS+jXgjYI3QIHJgIH0KIFdRC8DhYNBBahTgv6EGcEUZURSEuBtKy+Rdq/NF6FZ+Kj0eSP97H5uFt/4y9GA721POPJpAAAAABJRU5ErkJggg==)
> ```
> 
> Does that suit your need?

This is what I have done so far, and it works perfectly, (with some other extensions that allow pasting from clipboard in base64).
However, changing an image will be more difficult, than just swapping a file.

Thank you for the response.

**Comment by manuth** [comment]
So, basically the goal you're trying to reach is to have a single `.html`-file which contains all pictures rather than a diectory containing said pictures (like, for example, the `SelfContainedHTML`-option does)?

**Comment by Postur** [comment]
That is correct.

**Comment by Postur** [comment]
implemented #104  and tested.
Works flawlessly.
